### PR TITLE
Support product-registered commands

### PR DESCRIPTION
**Pull request by trilopin**
#### Summary
Support for slash commands in the context of Multi Product Architecture.

This PR has been extracted from the PR done to add [MPA support for playbooks](https://github.com/mattermost/mattermost-server/pull/22371) to have a cleaner merge path with the monorepo changes in mind.

The command will be executed in this order (higher ones can override lower ones):
1. plugins commands 
2. products commands 
3. custom commands 
4. builtin commands

The tests are a bit hacky when preparing the server, @isacikgoz kindly agreed to improve that approach 🥇.

#### Ticket Link
Nope

#### Release Note
```release-note
Added slash command support for products.
```

### CODE REVIEW

**Comment by lieut-data** [comment]
Thanks, @trilopin! @isacikgoz, these changes are blocking our ability to include Playbooks in the monorepo. Is the unit test coverage here sufficient to move us forward, or should we begin tracking the extra effort required?

**Comment by isacikgoz** [comment]
> Thanks, @trilopin! @isacikgoz, these changes are blocking our ability to include Playbooks in the monorepo. Is the unit test coverage here sufficient to move us forward, or should we begin tracking the extra effort required?

I think coverage-wise the approach is okay. I believe there is a room for improvement in the test helpers to be able to safely pass a mock product. We can defer that improvement, I'll create a ticket so that we can tackle later. There is no need to block monorepo+mpa/playbooks efforts.

File: app/plugin_commands.go [hunk 1]
```diff
@@ -102,16 +102,25 @@ func (ch *Channels) unregisterPluginCommands(pluginID string) {
 	ch.pluginCommands = remaining
 }
 
-func (a *App) PluginCommandsForTeam(teamID string) []*model.Command {
+// CommandsForTeam returns all the plugin and product commands for the given team.
+func (a *App) CommandsForTeam(teamID string) []*model.Command {
+	var commands []*model.Command
+
 	a.ch.pluginCommandsLock.RLock()
 	defer a.ch.pluginCommandsLock.RUnlock()
```
Comment (isacikgoz) [inline app/plugin_commands.go:110]: No need to use `defer`, we can simply unlock after the range for `a.ch.pluginCommands`.

File: app/plugin_commands_test.go [hunk 2]
```diff
@@ -439,3 +443,188 @@ func TestPluginCommand(t *testing.T) {
...
+			Dependencies: map[product.ServiceKey]struct{}{},
+		},
+	}
+
+	t.Run("Execute product command", func(t *testing.T) {
+		th := Setup(t).InitBasic()
+		defer th.TearDown()
+		// Server hijack.
+		// This must be done in a cleaner way.
+		th.Server.initializeProducts(products, th.Server.services)
```
Comment (isacikgoz) [inline app/plugin_commands_test.go:481]: Let's add a TODO statement here if we agree deferring this to a later time. (with the ticket number)

**Review by isacikgoz — APPROVED** [review]
LGTM

**Comment by trilopin** [comment]
@isacikgoz let me know if I should ask review from anyone else, or any other steps before merging.

**Comment by isacikgoz** [comment]
@trilopin we don't need to wait for another, if @lieut-data is okay with the process, I'd say let's merge it :)

**Review by lieut-data — APPROVED** [review]
These changes make sense to me! One concern is @saturninoabril's requesting that all PRs have QA Review attached. Worst case, we can rebase this into Doug's monorepo branch if needed.

**Comment by trilopin** [comment]
> These changes make sense to me! One concern is @saturninoabril's requesting that all PRs have QA Review attached. Worst case, we can rebase this into Doug's monorepo branch if needed.

Note that without playbooks-mpa code, the qa here is limited to check if plugin/custom/builtin slash commands still work fine. But to test product slash commands, we'll need the mpa-playboks  branches. cc @saturninoabril

**Comment by trilopin** [comment]
@lieut-data @isacikgoz can I close this? I guess it doesn't make sense anymore now that monorepo is merged.

**Comment by lieut-data** [comment]
Yes! Thank you for making this possible into the mono-repo :)

**Pull request closed without merging**